core/src/consensus.rs

```diff
@@ -319,6 +336,88 @@ impl Tower {
+                let last_vote_ancestors = ancestors.get(&last_vote).unwrap();
+                let switch_slot_ancestors = ancestors.get(&switch_slot).unwrap();
+
+                if switch_slot_ancestors.contains(last_vote) {
+                    // If the `switch_slot is a descendant of the last vote,
+                    // no switching proof is neceessary
+                    return true;
+                }
+
+                // Should never consider switching to an ancestor
+                // of your last vote
+                assert!(!last_vote_ancestors.contains(&switch_slot));
+
+                let mut locked_out_stake = 0;
+                let mut locked_out_vote_accounts = HashSet::new();
+                for (candidate_slot, descendants) in descendants.iter() {
+                    // 1) Only consider lockouts a tips of forks as that
+                    //    includes all ancestors of that fork.
+                    // 2) Don't consider lockouts on the `last_vote` itself
+                    //    or any of its descendants
+                    if !descendants.is_empty()
+                        || candidate_slot == last_vote
+                        || ancestors.get(&candidate_slot).unwrap().contains(last_vote)
+                    {
+                        continue;
+                    }
+
+                    // By the time we reach here, any ancestors of the `last_vote`,
+                    // should have been filtered out, as they all have a descendant,
+                    // namely the `last_vote` itself.
+                    assert!(!last_vote_ancestors.contains(candidate_slot));
+
+                    // Evaluate which vote accounts in the bank are locked out
+                    // in the interval candidate_slot..last_vote, which means
+                    // finding any lockout intervals in the `lockout_intervals` tree
+                    // for this bank that contain `last_vote`.
+                    let lockout_intervals = &progress
+                        .get(&candidate_slot)
+                        .unwrap()
+                        .fork_stats
+                        .lockout_intervals;
+                    // Find any locked out intervals in this bank with endpoint >= last_vote,
+                    // implies they are locked out at last_vote
+                    for (_, value) in lockout_intervals.range((Included(last_vote), Unbounded)) {
+                        for (lockout_interval_start, vote_account_pubkey) in value {
+                            // Only count lockouts on slots that are:
+                            // 1) Not ancestors of `last_vote`
+                            // 2) Not from before the current root as we can't determine if
+                            // anything before the root was an ancestor of `last_vote` or not
+                            if !last_vote_ancestors.contains(lockout_interval_start)
+                                && ancestors.contains_key(lockout_interval_start)
```

Comment:
@aeyakovenko, I'm filtering out any branches from before the root, so those forks can't be included in the switching proofs, even though they may be locked out above our last vote. I don't think it's a huge issue, but I could be missing something...